OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -152,31 +183,20 @@ bool TracerThread::OpenUprobes(const std::vector<int32_t>& cpus) {
       uprobes_uretprobes_ids_to_function_.emplace(stream_id, &function);
       uretprobes_ids_.insert(stream_id);
     }
+  }
 
-    // Redirect all uprobes and uretprobes on the same cpu to a single ring
-    // buffer to reduce the number of ring buffers.
-    for (int32_t cpu : cpus) {
-      int uprobes_fd = function_uprobes_fds_per_cpu.at(cpu);
-      int uretprobes_fd = function_uretprobes_fds_per_cpu.at(cpu);
-      if (uprobes_ring_buffer_fds_per_cpu.contains(cpu)) {
-        // Redirect to the already opened ring buffer.
-        int ring_buffer_fd = uprobes_ring_buffer_fds_per_cpu.at(cpu);
-        perf_event_redirect(uprobes_fd, ring_buffer_fd);
-        perf_event_redirect(uretprobes_fd, ring_buffer_fd);
-      } else {
-        // No ring buffer has yet been created for this cpu, as this is the
-        // first uprobes to have been opened successfully. Hence, create a
-        // ring buffer for this cpu associated to uprobes_fd and redirect the
-        // uretprobes to it. The other uprobes and uretprobes for this cpu
-        // will be redirected to this ring buffer.
-        int ring_buffer_fd = uprobes_fd;
-        std::string buffer_name = absl::StrFormat("uprobes_uretprobes_%d", cpu);
-        ring_buffers_.emplace_back(ring_buffer_fd, UPROBES_RING_BUFFER_SIZE_KB,
-                                   buffer_name);
-        uprobes_ring_buffer_fds_per_cpu[cpu] = ring_buffer_fd;
-        // Must be called after the ring buffer has been opened.
-        perf_event_redirect(uretprobes_fd, ring_buffer_fd);
-      }
+  for (auto& [/*int32_t*/ cpu, /*std::vector<int>*/ fds] : fds_per_cpu) {
```

Comment:
The reason for changing this code is because you don't always open both the uprobe and the uretprobe, right? But as I said in the previous comment, I think the behavior on errors is different now (I might be wrong though). It can be made the same, but it needs additional/[even more] bookkeeping of file descriptors. That's why it was implemented that way. Wouldn't it be better to add `if`s based on `open_uprobe`\`open_uretprobe` to the old code?

Reply:
Please see new implementation.